hazelcast/src/test/java/com/hazelcast/map/impl/query/LocalKeySetTest.java

```diff
@@ -0,0 +1,121 @@
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Map;
+import java.util.Set;
+
+import static com.hazelcast.test.TestCollectionUtils.setOf;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class LocalKeySetTest extends HazelcastTestSupport {
+    private HazelcastInstance local;
+    private IMap<String, String> map;
+    private SerializationService serializationService;
+    private HazelcastInstance remote;
+    private String remoteKey1;
+    private String remoteKey2;
+    private String remoteKey3;
+    private String localKey1;
+    private String localKey2;
+    private String localKey3;
+
+    @Before
+    public void setup() {
+        HazelcastInstance[] cluster = createHazelcastInstanceFactory(2).newInstances();
+        local = cluster[0];
+        remote = cluster[1];
+        map = local.getMap(randomName());
+        serializationService = getSerializationService(local);
+
+        remoteKey1 = generateKeyOwnedBy(remote);
+        remoteKey2 = generateKeyOwnedBy(remote);
+        remoteKey3 = generateKeyOwnedBy(remote);
+
+        localKey1 = generateKeyOwnedBy(local);
+        localKey2 = generateKeyOwnedBy(local);
+        localKey3 = generateKeyOwnedBy(local);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void whenPredicateNull() {
+        map.localKeySet(null);
+    }
+
+    @Test
+    public void whenMapEmpty() {
+        Set<String> result = map.localKeySet(TruePredicate.INSTANCE);
+
+        assertTrue(result.isEmpty());
+    }
+
+    @Test
+    public void whenSelecting_withoutPredicate() {
+        map.put(localKey1, "a");
+        map.put(localKey2, "b");
+        map.put(remoteKey1, "c");
+
+        Set<String> result = map.localKeySet();
+
+        assertEquals(setOf(localKey1, localKey2), result);
+    }
+
+    @Test
+    public void whenSelectingAllEntries() {
+        map.put(localKey1, "a");
+        map.put(localKey2, "b");
+        map.put(remoteKey1, "c");
+
+        Set<String> result = map.localKeySet(TruePredicate.INSTANCE);
+
+        assertEquals(setOf(localKey1, localKey2), result);
+    }
+
+    @Test
+    public void whenSelectingSomeEntries() {
+        map.put(localKey1, "good1");
+        map.put(localKey2, "bad");
+        map.put(localKey3, "good2");
+        map.put(remoteKey1, "good3");
+        map.put(remoteKey2, "bad");
+        map.put(remoteKey3, "good4");
+
+        Set<String> result = map.localKeySet(new GoodPredicate());
+
+        assertEquals(setOf(localKey1, localKey3), result);
+    }
+
+    @Test
+    public void testResultType() {
```

Comment:
Since there is no more need to send more information than needed, perhaps we can use the same approach as client. On the client the whole collection is fully deserialized and returned. So you get a regular collection with real objects. Instead of this intercepted thing.